docs/inventory_plugins.md

```diff
@@ -0,0 +1,181 @@
+An effort was made to keep imports working in the exact same way after
+the switchover. However, the inventory plugins are a fundamental rewrite
+and many elements of default behavior has changed. Because of that,
+a `compatibility_mode` toggle was added. This defaults to True.
+
+Turning off compatibility mode will be more future-proof.
+Keeping it on, will be more stable and consistent.
+
+### Changes with Compatibility Mode Off
+
+The set of `hostvars` will be almost completely different, using new names
+for data which is mostly the same content. You can see the jinja2 keyed_groups
+construction used in compatibility mode to help get a sense of what
+new names replace old names.
+
+If you turn compatibility mode off or downgrade Ansible, you should
+consider turning on `overwrite` and `overwrite_vars` to get rid of stale
+variables (and potentially groups) no longer returned by the import.
+
+In many cases, the host names will change. In all cases, accurate host
+tracking will still be maintained via the host `instance_id`.
+(after: https://github.com/ansible/awx/pull/3362)
+
+Group names will be sanitized with compatibility mode turned off.
+That means that characters such as "-" will
+be replaced by underscores "\_". In some cases, this means that a large
+fraction of groups get renamed as you move from scripts to plugins.
+This will become the default Ansible behavior on the CLI eventually.
+
+### Changes with Compatibility Mode On
+
+Programatically-generated examples of inventory file syntax used in
+updates (with dummy data) can be found in `awx/main/tests/data/inventory/scripts`,
+these demonstrate the inventory file syntax used to restore old behavior
+from the inventory scripts.
+
+#### hostvar keys and values
+
+More hostvars will appear if the inventory plugins are used with compatibility
+mode on. To maintain backward compatibility,
+the old names are added back where they have the same meaning as a
+variable returned by the plugin. New names are not removed.
+
+Some hostvars will be lost, because of general deprecation needs.
+
+ - ec2, see https://github.com/ansible/ansible/issues/52358
+ - gce (see https://github.com/ansible/ansible/issues/51884)
+   - `gce_uuid` this came from libcloud and isn't a true GCP field
+     inventory plugins have moved away from libcloud
+
+ The syntax of some hostvars, for some values, will change.
+
+  - ec2
+    - old: "ec2_block_devices": {"sda1": "vol-xxxxxx"}
+    - new: "ec2_block_devices": {"/dev/sda1": "vol-xxxxxx"}
+
+#### Host names
+
+Host names might change, but tracking host identity via `instance_id`
+will still be reliable.
+
+## How do I write my own Inventory File?
+
+If you do not want any of this compatibility-related functionality, then
+you can add an SCM inventory source that points to your own file.
+You can also apply a credential of a `managed_by_tower` type to that inventory
+source that matches the credential you are using, as long as that is
+not `gce` or `openstack`.
+
```

Comment:
Eventually, we will need docs on project-included inventory plugin usage.

Reply:
> project-included inventory plugin usage.

 I'm not entirely sure I follow what use case this is denoting.

If you want to bring your own `aws_ec2.yml` file via source control, that's what this is about.

If you want to bring your own `openshift.yml` file, then create your own custom credential type, and proceed.

If you want to bring your own `cow.yml` file, see https://github.com/ansible/test-playbooks/pull/52 for examples. See https://github.com/ansible/ansible/pull/51177, that is a blocking issue, preventing users from being able to do this, which has been established for a while. I have been working on that use case, but until the core fix is in, there is nothing for us to do.

Reply:
Yes, that's what I mean. Once https://github.com/ansible/ansible/pull/51177, we'll need to document how to bring your own entirely custom plugin (as opposed to a configuration for an existing plugin) to the party.

Reply:
Also tracking in https://github.com/ansible/awx/issues/3113